legate/core/runtime.py

```diff
@@ -711,14 +729,389 @@ def record_partition(self, index_space, functor, index_partition):
         self._index_partitions[key] = index_partition
 
 
+class FusionChecker(object):
+    def __init__(self, ops, contexts, runtime):
+        """
+        This is a class containing a list of constraints for fusing ops
+        It emits whether or not a given list of ops can be fused
+        """
+        self.constraints = []
+        self.ops = ops
+        self.contexts = contexts
+        self.runtime=runtime
+        self.partitioners = []
+        self.strategies = []
+
+    def register_constraint(self, fusion_constraint_rule):
+        self.constraints.append(fusion_constraint_rule)
+
+    def supress_small_fusions(self, intervals, threshold):
+        #find if there's a fusable sub window of length
+        #greater than or equal to fusion_thresh
+        final_set = []
+        fusable=False
+        for interval in intervals:
+            if interval[1] - interval[0]  >=threshold:
+                final_set.append(interval)
+                fusable = True
+            else:
+                for i in range(interval[0], interval[1]):
+                    final_set.append((i, i+1))
+        return fusable, final_set
+
+    def can_fuse(self):
+        #starttime = datetime.datetime.now()
+        for op in reversed(self.ops):
+            must_be_single = any(len(gop.scalar_outputs) > 0 for gop in [op])
+            partitioner = Partitioner(self.runtime, [op], must_be_single=must_be_single)
+            self.partitioners.append( partitioner )
+            strategy = partitioner.partition_stores()
+            for output, part, in zip(op._outputs, op._output_parts):
+                partition = strategy.get_partition(part)
+                output.set_key_partition(partition)
+                key_part = partition
+                for input in op._inputs:
+                    if input.shape==output.shape:
+                        input.set_key_partition(key_part)
+            #if len(op.inputs)>1:
```

Comment:
This part is important. It sets key partitions so the partition of a store in fused op can propagate  to other references of the same store.
There may be a better way to do this than 
```               for input in op._inputs:
                    if input.shape==output.shape:
                        input.set_key_partition(key_part                        ```

Reply:
TODO: use the alignment data structure to check if the input should be aligned with the output

Reply:
edit: changed to use alignment/constraint info, will be pushed